ra/ra.go

```diff
@@ -996,60 +1009,85 @@ func (ra *RegistrationAuthorityImpl) FinalizeOrder(ctx context.Context, req *rap
 		return nil, err
 	}
 
-	// Step 3: Issue the Certificate
-	cert, err := ra.issueCertificateInner(
-		ctx, csr, accountID(req.Order.RegistrationID), orderID(req.Order.Id))
-
-	// Step 4: Fail the order if necessary, and update metrics and log fields
-	var result string
+	// Update the order status locally since the SA doesn't return the updated
+	// order itself after setting the status
 	order := req.Order
-	if err != nil {
-		// The problem is computed using `web.ProblemDetailsForError`, the same
-		// function the WFE uses to convert between `berrors` and problems. This
-		// will turn normal expected berrors like berrors.UnauthorizedError into the
-		// correct `urn:ietf:params:acme:error:unauthorized` problem while not
-		// letting anything like a server internal error through with sensitive
-		// info.
-		ra.failOrder(ctx, req.Order, web.ProblemDetailsForError(err, "Error finalizing order"))
-
-		// Update the order status locally since the SA doesn't return the updated
-		// order itself after setting the status
-		order.Status = string(core.StatusInvalid)
-
-		logEvent.Error = err.Error()
-		beeline.AddFieldToTrace(ctx, "issuance.error", err)
-		result = "error"
-	} else {
-		// Update the order status locally since the SA doesn't return the updated
-		// order itself after setting the status
-		order.CertificateSerial = core.SerialToString(cert.SerialNumber)
-		order.Status = string(core.StatusValid)
+	order.Status = string(core.StatusProcessing)
+
+	// We do steps 3 and 4 in a goroutine so that we can better handle latency
+	// from getting SCTs and writing the (pre)certificate to the database. This
+	// lets us return the order in the Processing state to the client immediately,
+	// prompting them to poll the Order object and wait for it to be put into its
+	// final state.
+	//
+	// We track this goroutine's lifetime in two separate waitgroups: one on the
+	// RA itself, so that it can wait for all goroutines to drain during shutdown,
+	// and one local to this request, so that this function can wait on the
+	// goroutine if the AsyncFinalize flag isn't enabled.
+	ra.finalizeCounter.Add(1)
+	var wg sync.WaitGroup
+	wg.Add(1)
+	go func() {
+		// Step 3: Issue the Certificate
+		var cert *x509.Certificate
+		cert, err = ra.issueCertificateInner(
```

Comment:
As long as it's still synchronous, the updated `order` object is also an output, not just the single `err`.

I did have this as a separate function for a little bit (`finalizeOrderInner`) but specifically re-inlined it because the function only actually does two things: call issueCertificateInner, and handle its return value. It can't get moved into issueCertificateInner, because then all of the result handling would have to be replicated in multiple places. And making it its own function that only does one function call felt like re-adding the many difficult-to-navigate layers of abstraction that were just simplified in #6588. As long as this is operating in synchronous mode, it can just be read as straight-line code.

Maybe we leave this simplification for when the AsyncFinalize flag is removed, and we no longer have to support both sync and async flows?

Reply:
Long inline goroutines are scary for two reasons:

1. They are not explicit about which variables they capture (and possibly mutate). This is great for short goroutines, since it reduces duplication. But for long goroutines it is hard to figure out with accuracy. As an example, I missed that `order` is also an output parameter (or more precisely an in/out parameter that gets mutated).

2. They run concurrently with the function from which they capture variables. That means it's really important to check that access to those variables is properly synchronized. But it's hard to do that because it's not clear which variables need synchronization.

Actually, come to think of it - isn't there a data race here when the AsyncFinalize feature flag is enabled? The parent function returns `order`, which is concurrently being mutated by the goroutine.

Reply:
I've now moved this into a very thin helper function that takes ctx, order, csr, and logEvent as arguments, and returns the updated order and an error (which are only used in the synchronous case).